sdk/core/azure-core/src/main/java/com/azure/core/implementation/jackson/ObjectMapperShim.java

```diff
@@ -0,0 +1,397 @@
+
+        /*
+         * Begin by looping over all declared fields and initializing all header collection information.
+         */
+        for (final Field declaredField : declaredFields) {
+            if (!declaredField.isAnnotationPresent(HeaderCollection.class)) {
+                continue;
+            }
+
+            final Type declaredFieldType = declaredField.getGenericType();
+            if (!TypeUtil.isTypeOrSubTypeOf(declaredField.getType(), Map.class)) {
+                continue;
+            }
+
+            final Type[] mapTypeArguments = TypeUtil.getTypeArguments(declaredFieldType);
+            if (mapTypeArguments.length != 2
+                || mapTypeArguments[0] != String.class
+                || mapTypeArguments[1] != String.class) {
+                continue;
+            }
+
+            final HeaderCollection headerCollectionAnnotation = declaredField.getAnnotation(HeaderCollection.class);
+            final String headerCollectionPrefix = headerCollectionAnnotation.value().toLowerCase(Locale.ROOT);
+            final int headerCollectionPrefixLength = headerCollectionPrefix.length();
+            if (headerCollectionPrefixLength == 0) {
+                continue;
+            }
+
+            headerCollectionHandlers.add(new HeaderCollectionHandler(headerCollectionPrefix, declaredField));
+            headerCollectionsFirstCharacters.add(headerCollectionPrefix.charAt(0));
+        }
+
+        /*
+         * Then loop over all headers and check if they begin with any of the prefixes found.
+         */
+        for (final HttpHeader header : headers) {
+            String headerNameLower = header.getName().toLowerCase(Locale.ROOT);
+
+            /*
+             * Optimization to skip this header as it doesn't begin with any character starting header collections in
+             * the deserialized headers type.
+             */
+            if (!headerCollectionsFirstCharacters.contains(headerNameLower.charAt(0))) {
+                continue;
+            }
+
+            for (HeaderCollectionHandler headerCollectionHandler : headerCollectionHandlers) {
+                if (headerCollectionHandler.headerStartsWithPrefix(headerNameLower)) {
+                    headerCollectionHandler.addHeader(header.getName(), header.getValue());
+                }
+            }
+        }
+
+        /*
+         * Finally, inject all found header collection values into the deserialized headers.
+         */
+        headerCollectionHandlers.forEach(h -> h.injectValuesIntoDeclaringField(deserializedHeaders, LOGGER));
+
+        return deserializedHeaders;
+    }
+
+    /*
+     * Helper method that gets the value for the given key from the cache.
+     */
+    static <K, V> V getFromCache(K key, Map<K, V> cache, Function<K, V> compute) {
+        if (cache.size() >= CACHE_SIZE_LIMIT) {
+            cache.clear();
+        }
+
+        return cache.computeIfAbsent(key, compute);
```

Comment:
Before clearing the cache, should we check to see if the entry exists and clearing may not be required?

Reply:
still, I'd like to preserve existing behavior since this is not part of this change.